xhydro/frequency_analysis/local.py

```diff
@@ -0,0 +1,937 @@
+                return max_ds
+            else:
+                # TODO add year from grouped_ds.data.dt.year
+                # and make full str start_date and end_date
+                grouped_ds.data.coords["start_date"] = pd.to_datetime(
+                    str(season_vals[0]), format="%j"
+                ).strftime("%m-%d")
+                grouped_ds.data.coords["end_date"] = pd.to_datetime(
+                    str(season_vals[1]), format="%j"
+                ).strftime("%m-%d")
+
+                return (
+                    grouped_ds.custom_group_by(season_vals[0], season_vals[1])
+                    .max()
+                    .where(
+                        grouped_ds.get_bool_over_tolerence(tolerence, season), drop=True
+                    )
+                )
+
+        if seasons:
+            # Creating a new dimension of season and
+            # merging all Dataset from max_over_one_season
+            ds = xr.concat(
+                [
+                    max_over_one_season(grouped_ds, tolerence, season)
+                    .assign_coords(season=season)
+                    .expand_dims("season")
+                    for season in seasons
+                ],
+                dim="season",
+            )
+            return ds[list(ds.keys())[0]]
+
+        else:
+            # TODO Tolerence not used if no period is defined
+            return (
+                grouped_ds.data.groupby("time.year")
+                .max()
+                .assign_coords(season="Whole year")
+                .expand_dims("season")
+            )
+
+    def calculate_volume(self, dates: Union[list, xr.Dataset] = None, tolerence=0.15):
+        ds = self.copy()
+
+        def conversion_factor_to_hm3(timestep):
+            # flow is in m³/s and we want m³, so we multiply by seconds
+            # TODO check if last date is included
+            return pd.to_timedelta(1, unit=timestep).total_seconds()
+
+        if dates is None:
+            # TODO use season dates
+            pass
+        elif isinstance(dates, list):
+            # TODO bool over tolerence takes season, generalise
+            with warnings.catch_warnings():  # Removes overlaping warning
+                warnings.simplefilter("ignore")
+                self.season = ["Volumes", dates[0], dates[1]]
+            grouped_ds = (
+                ds.custom_group_by(dates[0], dates[1])
+                .sum()
+                .where(ds.get_bool_over_tolerence(tolerence, "Volumes"), drop=True)
+            )
+            self.rm_season("Volumes")
+            # Transform tp hm³
+            # TODO add start and end and clear other attributes
+            grouped_ds = (
+                grouped_ds
+                * xr.apply_ufunc(
+                    conversion_factor_to_hm3,
+                    grouped_ds["timestep"],
+                    input_core_dims=[[]],
+                    vectorize=True,
+                )
+                * (dates[1] - dates[0])
```

Comment:
As it is written right now, I think that you basically sum twice. The first time is here (and would result in biased results if there are NaNs, so .mean() is more robust), and the second time is at Line 522 when you multiply the results by the number of days.

```suggestion
                .mean()
                .where(ds.get_bool_over_tolerence(tolerence, "Volumes"), drop=True)
            )
            self.rm_season("Volumes")
            # Transform tp hm³
            # TODO add start and end and clear other attributes
            grouped_ds = (
                grouped_ds
                * xr.apply_ufunc(
                    conversion_factor_to_hm3,
                    grouped_ds["timestep"],
                    input_core_dims=[[]],
                    vectorize=True,
                )
                * (dates[1] - dates[0] + 1)
```

Reply:
Let's say I have: `ds2.streamflow.isel(time=slice(0,5)).values`
`array([[4.56, 4.25, 3.99, 3.77, 3.62]], dtype=float32)`

Daily volumes: `ds2.streamflow.isel(time=slice(0,5)).values * 60*60*24`
`array([[393984., 367200., 344736., 325728., 312768.]], dtype=float32)`

Sum: `ds2.streamflow.isel(time=slice(0,5)).values * 60*60*24).sum()`
`1744416.0 m³`, or `1.744416 hm³`. This should be the right answer, no?

However, `ds_fa_vol = ds_fa.calculate_volume(tolerence=0.15, dates=[1, 5])`
`6.97766418`, which is `1.744416 * (5 - 1)`

Reply:
@TC-FF Qu'en penses-tu ?

Reply:
Est-ce qu'il n'y a pas un danger avec la combinaison de `.sum()` et `tolerance` ? Si tu as 15% de données manquantes pour une année donnée, la somme risque d'être pas mal biaisée.

Reply:
Pour la tolérance, ça me semble ok, comme c'est par année, si on dépasse, l'année n'est pas inclue. Mais en effet si l'on est sous la tolérance, est-ce qu'on veut pondérer selon le nb de valeurs manquante ? Dans un cas de datas fixes, je dirais oui, dans un cas de dates variable, ¸a me semble moins nécessaire.